README.md
# Magento 2 Image Optimizer

This Magento 2 module is a wrapper based on the package [Spatie Image optimizer](https://github.com/spatie/image-optimizer).

## Installation
- `composer require justbetter/magento2-image-optimizer`
- `bin/magento module:enable JustBetter_ImageOptimizer`
- `bin/magento setup:upgrade && bin/magento setup:static-content:deploy`

### Optimize all images in console
- Run `bin/magento justbetter:imageoptimizer:optimizeall` to resize all images in the media folder.
- Additional parameters (will overwrite any option in the magento config):
    - `--jpg_compression="85"`
    - `--jpg_strip_all="1"`
    - `--jpg_all_progressive="0"`
    - `--png_quality_min_max="10-15"`
    - `--png_interlace="0"`
    - `--png_optimization_level="8"`

### Optimization tools

The package will use these optimizers if they are present on your system:

- [JpegOptim](http://freecode.com/projects/jpegoptim)
- [Optipng](http://optipng.sourceforge.net/)
- [Pngquant 2](https://pngquant.org/)
- [SVGO](https://github.com/svg/svgo)
- [Gifsicle](http://www.lcdf.org/gifsicle/)

Here's how to install all the optimizers on Ubuntu:

```bash
sudo apt-get install jpegoptim
sudo apt-get install optipng
sudo apt-get install pngquant
sudo npm install -g svgo
sudo apt-get install gifsicle
```

And here's how to install the binaries on MacOS (using [Homebrew](https://brew.sh/)):

```bash
brew install jpegoptim
brew install optipng
brew install pngquant
brew install svgo
brew install gifsicle
```

## Configuration
- Options for the module are defined in the backend under Stores > Configuration > JustBetter > Image optimizer configuration.
- Possible options:
    - Log the compression in the system.log
    - Change compression of jpg files
    - Change compression of png files

## Compatibility
The module is tested on magento version 2.2.x with Spatie image optimizer version 1.0.x

## Ideas, bugs or suggestions?
Please create a [issue](https://github.com/justbetter/magento2-image-optimizer/issues) or a [pull request](https://github.com/justbetter/magento2-image-optimizer/pulls).

## Todo
- ~~Configurable options for compression~~
- ~~Compress all library images in console command~~

## License
[MIT](LICENSE)

---

<a href="https://justbetter.nl" title="JustBetter"><img src="https://raw.githubusercontent.com/justbetter/art/master/justbetter-logo.png" width="200px" alt="JustBetter logo"></a>


CHANGELOG.md
CHANGELOG

2018-10-17 - 1.2.0

    * Added options in magento backend to set compression ratio
    * Added cli options in OptimizeImageCommand

2018-04-30 - 1.1.0

    * Added console command to optimize all images in magento 2 (media dir)

2018-04-06 - 1.0.0

    * Initial commit module - basic working module


Helper/Data.php
<?php

namespace JustBetter\ImageOptimizer\Helper;

use Magento\Store\Model\ScopeInterface;
use Spatie\ImageOptimizer\OptimizerChain;
use Magento\Framework\App\Helper\Context;
use Spatie\ImageOptimizer\Optimizers\Svgo;
use Spatie\ImageOptimizer\Optimizers\Optipng;
use Magento\Store\Model\StoreManagerInterface;
use Spatie\ImageOptimizer\Optimizers\Pngquant;
use Spatie\ImageOptimizer\Optimizers\Gifsicle;
use Spatie\ImageOptimizer\Optimizers\Jpegoptim;
use Magento\Framework\App\Helper\AbstractHelper;

/**
 * Class Data
 *
 * @package JustBetter\Sentry\Helper
 */
class Data extends AbstractHelper
{
    protected $configPaths = [
        'image_optimizer/general/',
        'image_optimizer/jpg/',
        'image_optimizer/png/',
    ];

    /**
     * @var StoreManagerInterface
     */
    protected $storeManager;

    /**
     * @var array
     */
    protected $configKeys = [
        'log',
        'enabled',
        'jpg_compression',
        'jpg_strip_all',
        'jpg_all_progressive',
        'png_quality_min_max',
        'png_interlace',
        'png_optimization_level',
    ];

    public $cliConfigKeys = [
        'jpg_compression' => '85',
        'jpg_strip_all' => true,
        'jpg_all_progressive' => true,
        'png_quality_min_max' => '65-80',
        'png_interlace' => true,
        'png_optimization_level' => 2,
    ];

    /**
     * @var array
     */
    protected $config = [];

    /**
     * @var State
     */
    protected $appState;

    /**
     * Data constructor.
     *
     * @param Context               $context
     * @param StoreManagerInterface $storeManager
     */
    public function __construct(
        Context $context,
        StoreManagerInterface $storeManager
    ) {
        $this->storeManager = $storeManager;

        parent::__construct($context);
    }

    /**
     * @param      $field
     * @param null $storeId
     * @return mixed
     */
    public function getConfigValue($field, $storeId = null)
    {
        return $this->scopeConfig->getValue(
            $field,
            ScopeInterface::SCOPE_STORE,
            $storeId
        );
    }

    /**
     * @param      $code
     * @param null $storeId
     * @return mixed
     */
    public function getGeneralConfig($code, $path, $storeId = null)
    {
        return $this->getConfigValue($path . $code, $storeId);
    }

    /**
     * @return array
     */
    public function collectModuleConfig()
    {
        foreach ($this->configPaths as $path) {
            foreach ($this->configKeys as $configKey) {
                $configValue = $this->getGeneralConfig($configKey, $path);
                if ($configValue !== null) {
                    $this->config[ $configKey ] = $configValue;
                }
            }
        }

        return $this->config;
    }

    /**
     * @return bool
     */
    public function isActive()
    {
        return (! empty($this->config) && array_key_exists('enabled', $this->config) && $this->config['enabled']);
    }

    /**
     * Create custom image optimizerChain
     * to set custom options
     *
     * @return Spatie\ImageOptimizer\OptimizerChain
     */
    public function customOptimizerChain($options = [])
    {
        // if options are not empty overwrite
        // with config values
        if (! empty($options)) {
            foreach($options as $name => $value) {
                if ($value) {
                    $this->config[$name] = $value;
                }
            }
        }

        return (new OptimizerChain)
            ->addOptimizer(new Jpegoptim([
                '-m'.$this->config['jpg_compression'],
                (bool)$this->config['jpg_strip_all'] ? '--strip-all' : '',
                (bool)$this->config['jpg_all_progressive'] ? '--all-progressive' : '',
            ]))

            ->addOptimizer(new Pngquant([
                $this->config['png_quality_min_max'] ? '--quality '.$this->config['png_quality_min_max'] : '',
                '--force',
            ]))

            ->addOptimizer(new Optipng([
                '-i'.(int)$this->config['png_interlace'],
                '-o'.(int)$this->config['png_optimization_level'],
                '-quiet',
            ]))

            ->addOptimizer(new Svgo([
                '--disable=cleanupIDs',
            ]))

            ->addOptimizer(new Gifsicle([
                '-b',
                '-O3',
            ]));
    }
}


Plugin/Media.php
<?php

namespace JustBetter\ImageOptimizer\Plugin;

use Psr\Log\LoggerInterface;
use JustBetter\ImageOptimizer\Helper\Data;
use Spatie\ImageOptimizer\OptimizerChainFactory;
use Magento\MediaStorage\Helper\File\Media as MagentoStorageMedia;

class Media
{
    /**
     * @var LoggerInterface
     */
    protected $logger;

    /**
     * @var Data
     */
    protected $imageOptimizerHelper;

    /**
     * @var array
     */
    protected $config;

    /**
     * Media constructor.
     *
     * @param LoggerInterface $logger
     * @param Data            $imageOptimizerHelper
     */
    public function __construct(
        LoggerInterface $logger,
        Data $imageOptimizerHelper
    ) {
        $this->imageOptimizerHelper = $imageOptimizerHelper;
        $this->config = $imageOptimizerHelper->collectModuleConfig();
        $this->logger= $logger;
    }

    /**
     * @param MagentoStorageMedia                     $media
     * @param                                         $mediaDirectory
     * @param                                         $path
     */
    public function beforeCollectFileInfo(MagentoStorageMedia $media, $mediaDirectory, $path)
    {
        if ($this->imageOptimizerHelper->isActive()) {
            $path = ltrim($path, '\\/');
            $fullPath = $mediaDirectory . '/' . $path;

            $optimizerChain = $this->imageOptimizerHelper->customOptimizerChain();

            if (array_key_exists('log', $this->config) && $this->config['log']) {
                $optimizerChain = $optimizerChain->useLogger($this->logger);
            }

            $optimizerChain->optimize($fullPath);
        }
    }
}


Console/Command/OptimizeImagesCommand.php
<?php

namespace JustBetter\ImageOptimizer\Console\Command;

use Exception;
use Psr\Log\LoggerInterface;
use RecursiveDirectoryIterator;
use Magento\Framework\Filesystem;
use Spatie\ImageOptimizer\OptimizerChain;
use JustBetter\ImageOptimizer\Helper\Data;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Input\InputInterface;
use Magento\Framework\App\Filesystem\DirectoryList;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * Class OptimizeImagesCommand
 *
 * @package JustBetter\ImageOptimizer\Console\Command
 */
class OptimizeImagesCommand extends Command
{
    protected $includeFiles = [
        'jpeg',
        'jpg',
        'png',
        'gif'
    ];

    protected $excludeDirs = [
        '(cache)',
        '(tmp)',
        '(.thumbs)'
    ];

    /**
     * @var Filesystem
     */
    protected $filesystem;

    /**
     * @var LoggerInterface
     */
    protected $logger;

    /**
     * @var Data
     */
    protected $imageOptimizerHelper;

    /**
     * @var array
     */
    protected $config;

    /**
     * OptimizeImagesCommand constructor.
     *
     * @param Filesystem      $filesystem
     * @param LoggerInterface $logger
     * @param Data            $imageOptimizerHelper
     */
    public function __construct(
        Filesystem $filesystem,
        LoggerInterface $logger,
        Data $imageOptimizerHelper
    ) {
        $this->filesystem = $filesystem;
        $this->imageOptimizerHelper = $imageOptimizerHelper;
        $this->config = $imageOptimizerHelper->collectModuleConfig();
        $this->logger= $logger;
        parent::__construct();
    }

    protected function configure()
    {
        /**
         *   Possible parameters to overwrite magento 2 options
         *  'compression' => 'Jpeg Compression (number 85)',
         *  'strip_all' => 'Jpeg Strip all (true or false)',
         *  'all_progressive' => 'Jpeg All progressive (true or false)',
         *  'quality_min_max' => 'PNG min and max quality (e.g. 65-80)',
         *  'interlace' => 'PNG interlace (true or false)',
         *  'optimization_level' => 'PNG optimization level (0 to 7)',
         */
        foreach ($this->imageOptimizerHelper->cliConfigKeys as $name => $value) {
            $options[] =  new InputOption(
                $name,
                null,
                InputOption::VALUE_OPTIONAL,
                $value
            );
        }

        $this->setName('justbetter:imageoptimizer:optimizeall')
            ->setDescription('Re-save all images with spatie package')
            ->setDefinition($options);

        parent::configure();
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $output->writeln('<info>Start Optimize all images</info>');

        $dir = new RecursiveDirectoryIterator(
            $this->filesystem->getDirectoryRead(DirectoryList::MEDIA)->getAbsolutePath(),
            RecursiveDirectoryIterator::SKIP_DOTS
        );

        $this->loopThrougDir($dir, $input, $output);
    }

    protected function loopThrougDir($dir, InputInterface $input, OutputInterface $output)
    {
        foreach ($dir as $fullPath => $fileinfo) {
            if ($fileinfo->isDir()) {
                $this->loopThrougDir(
                    new RecursiveDirectoryIterator(
                        $fileinfo->getPathName(),
                        RecursiveDirectoryIterator::SKIP_DOTS
                    ),
                    $input,
                    $output
                );
            }

            if (preg_match('~\.('.implode('|', $this->includeFiles).')$~', $fullPath) &&
                ! preg_match('/'.implode('|', $this->excludeDirs).'/', $fullPath)) {
                $this->optimizeImage($fullPath, $input, $output);
            }
        }
    }

    protected function optimizeImage($image, InputInterface $input, OutputInterface $output)
    {
        $overWriteOptions = [];

        foreach ($this->imageOptimizerHelper->cliConfigKeys as $name => $notused) {
            $overWriteOptions[$name] = $input->getOption($name);
        }

        $optimizerChain = $this->imageOptimizerHelper->customOptimizerChain($overWriteOptions);

        if (array_key_exists('log', $this->config) && $this->config['log']) {
            $optimizerChain = $optimizerChain->useLogger($this->logger);
        }

        try {
            $optimizerChain->optimize($image);
            $output->writeln('<info>Image: '.$image.' is optimized</info>');
        } catch (Exception $e) {
            $output->writeln('<error>Image not optimized, error: '.$e->getMessage());
        }
    }
}
